### ISSUE TITLE
ADOM creation of type "Fabric" not possible?

### ISSUE THREAD

**Issue by eJay57** [fortinet-ansible-dev/ansible-galaxy-fortimanager-collection#57]
Using fmgr_dvmdb_adom to create an ADOM is unable to create an ADOM of type "Fabric".

It seems that the parameter "restricted_prds" should be used to select the ADOM type and it is working well to create ADOM of type Fortigate, Fortiproxy, and so on... After testing all possible values, there seems to be no value allowing the creation of a "Fabric" type ADOM.
The documentation is not stating anything about "Fabric" type ADOMs.

Is there a way to create such ADOM with this module? If yes, it should be documented. If no, is it possible to add this feature?

**Comment by jpforcioli** [comment]
Hi,

You can use a playbook similar to this one:

```yaml
---
- hosts: 10.210.35.112
  connection: httpapi
  collections:
    - fortinet.fortimanager
  gather_facts: false
  vars:
    adom: fabric_adom_002
    fabric_adom_magic_number: 268435455
  tasks:
  - name: Add ADOM "{{ adom }}"
    fmgr_dvmdb_adom:
      bypass_validation: true
      state: present
      dvmdb_adom:
        name: "{{ adom }}"
        desc: Created with fortinet.fortimanager collection
        mr: 2
        os_ver: "7.0"
        restricted_prds: "{{ fabric_adom_magic_number }}"
    vars:
      ansible_command_timeout: 60
```

It is creating a Fabric ADOM 7.2 named `fabric_adom_002`

To avoid using this *magic* ADOM number, a specific `fabric` item will be created soon to be used with `restricted_prds` attribute (Reference: 823400).
However, to use this magic number, you also have to set the `bypass_validation` to `true` since it is not part of what our module expect in term of valid values for the `restricted_prds` attribute.

Note that creating a new ADOM takes some time on my FMG; I've to increase the default 30 sec **command** timeout to 60 sec.

Hope it helps.

**Comment by eJay57** [comment]
Hello,

thank you for the quick feedback.
I tested using the value you provided as `restricted_prds` and I confirm it works properly.

As this is just a workaround, is it possible to keep this issue open for tracking and close it when the `fabric` item is officially available?

**Comment by jpforcioli** [comment]
Internal Reference: 0823400

**Comment by cr0cdev** [comment]
Hi!
Any updates when Fabric will be available in `restricted_prds ` without magic number?

**Comment by dux-fortinet** [comment]
Hi @cr0cdev 

Thanks for the comment.
Please set "restricted_prds" as "fabric". It should work for all FMG versions released after 2023/01. (For now, please set bypass_validation to True, you will not need to do it in the future FMG Ansible Galaxy collection)

Example:
```
- name: Create adom
      fmgr_dvmdb_adom:
        bypass_validation: True
        state: present
        dvmdb_adom:
          desc: create through ansible
          name: ansible_fabric
          restricted_prds: fabric
```

**Comment by cr0cdev** [comment]
Hi!
I've tried with setting `restricted_prds` to fabric and `bypass_validation` to True.
Unfortunately it is still not working:
```
fatal: [fortimanager01]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "access_token": null,
            "bypass_validation": true,
            "dvmdb_adom": {
                "desc": "Created by Ansible",
                "flags": [
                    "no_vpn_console",
                    "per_device_wtp",
                    "per_device_fsw"
                ],
                "mode": "gms",
                "name": "ansible-test-2",
                "os_ver": 7.0,
                "restricted_prds": "fabric",
                "state": 1
            },
            "enable_log": false,
            "forticloud_access_token": null,
            "proposed_method": null,
            "rc_failed": null,
            "rc_succeeded": null,
            "state": "present",
            "workspace_locking_adom": null,
            "workspace_locking_timeout": 300
        }
    },
    "meta": {
        "request_url": "/dvmdb/adom",
        "response_code": -20002,
        "response_data": [],
        "response_message": "Invalid argument",
        "system_information": {
            "Admin Domain Configuration": "Enabled",
            "BIOS version": "04000002",
            "Branch Point": "1460",
            "Build": "1460",
            "Current Time": "Mon Nov 13 22:53:14 EET 2023",
            "Daylight Time Saving": "Yes",
            "FIPS Mode": "Disabled",
            "HA Mode": "Stand Alone",
            "Hostname": "FMG",
            "License Status": "Valid",
            "Major": 7,
            "Max Number of Admin Domains": 2,
            "Max Number of Device Groups": 3,
            "Minor": 2,
            "Offline Mode": "Disabled",
            "Patch": 4,
            "Platform Full Name": "FortiManager-VM64-HV",
            "Platform Type": "FMG-VM64-HV",
            "Release Version Information": " (GA)",
            "Serial Number": "FMG-XXXXXXXXXXX",
            "TZ": "Europe/Helsinki",
            "Time Zone": "(GMT+2:00) Helsinki, Riga,Tallinn.",
            "Version": "v7.2.4-build1460 230926 (GA)",
            "x86-64 Applications": "Yes"
        }
    },
    "rc": -20002
} 
```

**Comment by dux-fortinet** [comment]
It is because argument os_ver is not valid when you set "restricted_prds": "fabric".
Please do not specify os_ver: 7.0.

Following script works in my test environment:
```
- name: ADOM table, most attributes are read-only and can only be changed internally.
  fmgr_dvmdb_adom:
    bypass_validation: True
    state: present
    dvmdb_adom:
      desc: Created by Ansible
      name: ansible_fabric_1
      flags:
        - "no_vpn_console"
        - "per_device_wtp"
        - "per_device_fsw"
      restricted_prds: fabric
      state: 1
      mode: "gms"
```

**Comment by cr0cdev** [comment]
> It is because argument os_ver is not valid when you set "restricted_prds": "fabric". Please do not specify os_ver: 7.0.
> 
> Following script works in my test environment:
> 
> ```
> - name: ADOM table, most attributes are read-only and can only be changed internally.
>   fmgr_dvmdb_adom:
>     bypass_validation: True
>     state: present
>     dvmdb_adom:
>       desc: Created by Ansible
>       name: ansible_fabric_1
>       flags:
>         - "no_vpn_console"
>         - "per_device_wtp"
>         - "per_device_fsw"
>       restricted_prds: fabric
>       state: 1
>       mode: "gms"
> ```

After removing `os_ver` everything worked fine when `bypass_validation` is set to `true`.
Thanks dux for the help and explanation.

**Comment by dux-fortinet** [comment]
FortiManager Ansible Collection 2.4.0 is expected to be released on December 15th. This problem will be fixed at that time. Thanks for raising this issue.

**Comment by dux-fortinet** [comment]
Hi @cr0cdev, this problem has been fixed in FortiManager Ansible Collection 2.3.1
